### DISCUSSION TITLE
Correct way to autogenerate UUID primary key on server side, without a database roundtrip?

### DISCUSSION THREAD

**Discussion by AgarwalPragy** [sqlalchemy/sqlalchemy#10698] [Usage Questions] [answered]
When instantiating an ORM object, the `id` is `None`. The `id` isn't generated untill a `session.flush()` or `session.commit()` is performed.

```py
from uuid import UUID, uuid4
from sqlalchemy.ext.asyncio import AsyncAttrs
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

class ModelBase(AsyncAttrs, DeclarativeBase):
    __abstract__ = True
    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)

class Note(ModelBase):
    __tablename__ = 'notes'
    content: Mapped[str]

n = Note()
print(n.id)  # None
```

I would like to generate these uuids at the app server side, without waiting for a database round trip (hence the application of uuid instead of db generated autoincremented keys).

One way is to generate the id in `__init__`

```py
class Note(ModelBase):
    ...

    def __init__(self, **kwargs):
        kwargs['id'] = kwargs.get('id', uuid4())
        super().__init__(**kwargs)
```

Is this correct? I'm concerned that this might interfere with the inner workings of SQLAlchemy when it creates `Note` objects when loading row from the db.

**Comment by zzzeek** [comment, marked as answer]
You are asking for the uuid to be generated on the *client* side, not the *server* side.  The "server" here is your database.

the original mapping you are showing is still generating the uuid on the client side, however, SQLAlchemy isn't running the function until the INSERT statement proceeds.

If you want the uuid value to be present there right when you first make the object, your intuition is correct, do it in the constructor.  though I would avoid generating the uuid if one was passed:

```py
class Note(ModelBase):
    ...

    def __init__(self, **kwargs):
        if 'id' not in kwargs:
            kwargs['id'] = uuid4()
        super().__init__(**kwargs)
```

**Reply by AgarwalPragy** [reply]
Thanks.

**Reply by AgarwalPragy** [reply]
@zzzeek I guess this doesn't work when setting up a custom Base class? When I set the id like above, I get the following error when creating a new object
```py
TypeError: object.__init__() takes exactly one argument (the instance to initialize)
```


```py

class ModelBase(AsyncAttrs, DeclarativeBase):
    __abstract__ = True

    type_annotation_map = {
        int: BIGINT,
        datetime: TIMESTAMP(timezone=True),
        str: String,
        UUID: Uuid,
        bool: Boolean,
    }

    id: Mapped[UUID] = mapped_column(
        primary_key=True,
        # note: don't set default here because the value will be set by __init__
    )

    created_at: Mapped[datetime] = mapped_column(
        index=True,
        doc='Time of creation',
        server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        index=True,
        doc='Time of last modification',
        server_default=func.now(),
        onupdate=func.now())

    def __init__(self, **kwargs):
        if 'id' not in kwargs:
            # setting id here instead of in the column definition
            # because we want to generate it at object instantiation
            # instead of at flush time
            # https://github.com/sqlalchemy/sqlalchemy/discussions/10698
            kwargs['id'] = ulid.new().uuid
        super().__init__(**kwargs)

    @declared_attr
    @classmethod
    def __tablename__(cls) -> str:
        """Automatically generate table name from class name."""
        return classname_to_tablename(cls.__name__)
```

I'm encountering some really weird behavior.
1. Having no `__init__` in ModelBase works just fine (provided the `default` param is set for the primary key `id` column)
2. Adding a `__init__(self, **kwargs)` which calls `super().__init__(**kwargs)` throws the error `TypeError: object.__init__() takes exactly one argument (the instance to initialize)`
3. Adding a `__init__(self, **kwargs)` which calls `super().__init__()` (note that kwargs aren't passed) causes the object to be created without any data - thus later failing the not null constraints in the db

**Reply by AgarwalPragy** [reply]
> SQLAlchemy isn't running the function until the INSERT statement proceeds.

I'm trying to automatically create rows in another table by listening to the `before_flush` event. For the foreign_key constraint, the `id` of newly created rows is needed, which isn't generated by the time `before_flush` event listener is called even though the Column has a `default` attribute set.

Is this expected behavior? If yes, what is the correct way to set up the `id` for all models before the `before_flush` event is triggered?

**Reply by zzzeek** [reply]
question one:

the `DeclarativeBase` class does not have an `__init__` method.  The default `__init__` method is only on the base that you create.  so if you want to use the default `__init__`, you have to supply a separate class if you want to override it but still use it:

```py
class Base(DeclarativeBase):
    pass

class ModelBase(AsyncAttrs, Base):
    __abstract__ = True

    # ...

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

```

question two - if you are using your client side `uuid`, then you've populated that value when `__init__()` took place, you can use that.  otherwise, if this is some other kind of table with an auto-incrementing primary key value, you dont have access to that in `before_flush()`, unless you were using a `Sequence` and were manually incrementing it (which I don't recommend as Sequences are kind of legacy).   Instead, in order to add rows to a related table where the foreign key columns will automatically be populated with the value of the primary key after it's auto-incremented, use a [one to many relationship](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#one-to-many), appending related objects to the parent object.

Otherwise, if you absolutely can only use direct INSERT statements for these related rows, you would either   a. use INSERT for the primary rows also with RETURNING and do the whole thing manually, see https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#getting-new-objects-with-returning , or b. use the `after_flush()` event instead where the new unit-of-work objects you added will have their PKs and then you can run an INSERT on the connection for the related rows.

**Reply by AgarwalPragy** [reply]
Oh. That clears up everything! Thanks a ton for the quick answers :)

**Comment by jenstroeger** [comment]
Interesting, thanks for the conversation!

For a PG db server I’m using a [mixin](https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html):
```python
@declarative_mixin
class PKMixin:
    id: uuid.UUID = Column(UUID(as_uuid=True), primary_key=True, server_default=func.gen_random_uuid())
```
Based on the above conversation I thought I can overload the `__init__()` from this mixin to produce a UUID in case one wasn’t provided. But that doesn’t seem to work:
```python
@declarative_mixin
class PKMixin:
    def __init__(self, **kwargs): 
        if "id" not in kwargs:
            kwargs["id"] = uuid.uuid4()
        super().__init__(**kwargs)
    id: uuid.UUID = Column(UUID(as_uuid=True), primary_key=True, server_default=func.gen_random_uuid())

class Person(Base, PKMixin):
    __tablename__ = "person"
```
doesn’t call the mixin’s initializer whereas
```python
class Person(PKMixin, Base):
    __tablename__ = "person"
```
does. I’ve not checked that there’s a `super().__init__()` in every part of the call chain here but the only explanation for the behavior is that the initialization of the new instance is cut short? @zzzeek can you please shed some light on this behavior?

**Reply by axd1x8a** [reply]
I think it's because `DeclarativeBase` already defines `__init__` and as said [here](https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html):

> This works because Base here doesn’t define any of the variables that CommonMixin or HasLogRecord defines, i.e. `__tablename__`, `__table_args__`, `id`, etc. If the Base did define an attribute of the same name, the class placed first in the inherits list would determine which attribute is used on the newly defined class.

**Reply by zzzeek** [reply]
if you are calling `super()` from `class Person(X, Y)`, the call would start with `X` and `super()` would then call against `Y`.  If you reverse the order of the mixins then that would call `super()` on the `object` class.   if you want to force which class you call against, you can say `Y.__init__(self, **kwargs)` instead.

**Reply by jenstroeger** [reply]
There’s something I’m missing here 🤔

Looking at the [MRO](https://python-history.blogspot.com/2010/06/method-resolution-order.html) for a small test:
```python
>>> from sqlalchemy.orm import declarative_base
>>> Base = declarative_base()
>>> 
>>> from sqlalchemy.orm import declarative_mixin
>>> @declarative_mixin
... class Mixin:
...     def __init__(self, **kwargs):
...         print("Mixin init")
...         super().__init__(**kwargs)
... 
>>> class Person(Base, Mixin):
...     __tablename__ = "person"
...     pk = Column(Integer(), primary_key=True)
... 
>>> inspect.getmro(Base)
(<class 'sqlalchemy.orm.decl_api.Base'>, <class 'object'>)
>>> inspect.getmro(Mixin)
(<class '__main__.Mixin'>, <class 'object'>)
>>> inspect.getmro(Person)
(<class '__main__.Person'>, <class 'sqlalchemy.orm.decl_api.Base'>, <class '__main__.Mixin'>, <class 'object'>)
>>> Person()
<__main__.Person object at 0x106164280>
```
There’s no print from the `Mixin.__init__()` here. However, if I recreate that same class hierarchy:
```python
>>> class B:  # Hopefully standin for Base.
...     def __init__(self, *args, **kwargs):
...         print("B init")
...         super().__init__(*args, **kwargs)
... 
>>> class M:  # Standin for Mixin.
...     def __init__(self, *args, **kwargs):
...         print("M init")
...         super().__init__(*args, **kwargs)
... 
>>> class P(B, M):  # Standin for Person.
...     def __init__(self, *args, **kwargs):
...         print("X init")
...         super().__init__(*args, **kwargs)
... 
>>> inspect.getmro(P)
(<class '__main__.P'>, <class '__main__.B'>, <class '__main__.M'>, <class 'object'>)
>>> X()
P init
B init
M init
<__main__.P object at 0x106165780>
```
Here the `M.__init__()` indeed prints. So what’s different between the two, what am I missing?

**Reply by CaselIT** [reply]
The base default init does not call `super()`. It's this one https://github.com/sqlalchemy/sqlalchemy/blob/66be1482db06adb908432b2e3b41d9393d1319f7/lib/sqlalchemy/orm/decl_base.py#L2129-L2145

**Reply by jenstroeger** [reply]
> @CaselIT The base default init does not call `super()`.

Ah… and _that_ ends the call chain 👍🏼

**Comment by axd1x8a** [comment]
I tried what was suggested above, but went a bit further by calling functions to generate default values if they are not server-side:
```python
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.sql.functions import FunctionElement


class _Base(DeclarativeBase):
    ...


class Base(_Base):
    __abstract__ = True

    def __init__(self, **kwargs):
        for column in self.__table__.columns:
            if column.default is not None:
                if not column.default.is_scalar:
                    if isinstance(column.default.arg, FunctionElement):
                        continue
                    kwargs.setdefault(column.name, column.default.arg(None))
                else:
                    kwargs.setdefault(column.name, column.default.arg)
        super().__init__(**kwargs)
```
Solution is not the most elegant, but it works for my code, but I would like to hear recommendations from those more familiar with the inner workings of sqlalchemy.

Another problem with this solution is that records that depend on each other do not see id, and the entire transaction fails due to violation of foreign key constraints.

**Reply by jenstroeger** [reply]
Imagine your table has hundreds of columns (i.e. iterate over every single row) and you get/insert/update a lot of rows (i.e. create millions of objects) — this doesn’t seem to scale well?

**Reply by axd1x8a** [reply]
You may be right, but in my case there is no such thing planned, so there should be no performance issues, and I believe sqlachemy does the same anyway. The original idea was that the default values would be generated at initialization when it didn't require a database access, I always thought that was the case, but as it turns out it wasn't, which has now led to some problems when I wanted to reduce the number of database requests by only calling flush once.

**Reply by jenstroeger** [reply]
Depending on the db you can have the db generate defaults too, which is much faster than Python anyway — especially when you work with large data sets.

But you’re right, it doesn’t solve perfectly the problem of syncing Python objects with their respective db rows 🤷🏻‍♂️
